pkg/bearertoken/grpc.go

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package bearertoken
+
+import (
+	"context"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+const Key = "bearer.token"
+
+type tokenatedServerStream struct {
+	grpc.ServerStream
+	context context.Context
+}
+
+// getValidBearerToken attempts to retrieve the bearer token from the context.
+// It does not return an error if the token is missing.
+func getValidBearerToken(ctx context.Context, bearerHeader string) (string, error) {
+	bearerToken, ok := GetBearerToken(ctx)
+	if ok && bearerToken != "" {
+		return bearerToken, nil
+	}
+
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return "", nil
+	}
+
+	return tokenFromMetadata(md, bearerHeader)
+}
+
+// tokenFromMetadata extracts the bearer token from the metadata.
+func tokenFromMetadata(md metadata.MD, bearerHeader string) (string, error) {
+	bearerToken := md.Get(bearerHeader)
+	if len(bearerToken) < 1 {
+		return "", nil
+	}
+	if len(bearerToken) > 1 {
+		return "", nil
+	}
+	return bearerToken[0], nil
+}
+
+// NewStreamServerInterceptor creates a new stream interceptor that injects the bearer token into the context if available.
+func NewStreamServerInterceptor() grpc.StreamServerInterceptor {
+	return func(srv any, ss grpc.ServerStream, _ *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
```

Comment:
we can avoid the extra nesting
```suggestion
func NewStreamServerInterceptor() grpc.StreamServerInterceptor {
	return streamServerInterceptor
}

func streamServerInterceptor(srv any, ss grpc.ServerStream, _ *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
```